### PR TITLE
Add .gitignore file

### PR DESCRIPTION
**Pull request by davidcampelo**
Avoid "git status" to show annoying dirty working copy with generated files. :)

### CODE REVIEW

**Comment by avitase** [comment]
you should also add vim's *.swp files and the venv/ directory, for those using a virtualenv to install conan and cmake_format.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+**/CMakeFiles/
+**/CMakeCache.txt
+**cmake_install.cmake
+compile_commands.json
+Makefile
+**CTestTestfile.cmake
+
+# Visual Studio Code project files
+.vscode/
+language.settings.xml
```
Comment (aminya) [inline .gitignore:11]:
```suggestion
# IDE files
.vs/
.idea/
.vscode/
language.settings.xml
```

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+# Libary files
+lib/
+
+# Binary files
+imgui.ini
+bin/
+build/
+test/
+
```
Comment (aminya) [inline .gitignore:31]:
```suggestion
out/
```

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+language.settings.xml
+
+# Conan files
+conan.cmake
+conan.lock
+conanbuildinfo.cmake
+conanbuildinfo.txt
+conanfile.txt
+conaninfo.txt
+
```
Comment (aminya) [inline .gitignore:20]:
```suggestion
virtualenv/
```

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+graph_info.json
+
+# Libary files
+lib/
+
+# Binary files
+imgui.ini
+bin/
+build/
+test/
```
Comment (aminya) [inline .gitignore:30]: You should not gitignore the test folder!
```suggestion

```

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+compile_commands.json
+Makefile
+**CTestTestfile.cmake
+
+# Visual Studio Code project files
+.vscode/
+language.settings.xml
+
+# Conan files
+conan.cmake
```
Comment (aminya) [inline .gitignore:14]: You should not gitignore Conan.cmake. All the deps live here!!
```suggestion

```

Comment (davidcampelo) [reply .gitignore:14]: conan.cmake file is downloaded from the repo automatically the first time cmake runs. it's not part of the project.

Comment (aminya) [reply .gitignore:14]: There is another file called Conan.cmake under cmake folder. This conflicts with that.

Comment (davidcampelo) [reply .gitignore]: thanks for pointing that out. I believe using '/conan.cmake' would fix this.

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+.DS_Store
+.DS_Store?
+._*
+.Spotlight-V100
+.Trashes
+ehthumbs.db
+Thumbs.db
+.directory
+*~
+
```
Comment (aminya) [inline .gitignore:39]:
```suggestion
# IDE files
.vs/
.idea/
.vscode/
```

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+# Libary files
+lib/
+
+# Binary files
+imgui.ini
+bin/
+build/
+out/
+
```
Comment (aminya) [inline .gitignore:28]: Clion build folder
```suggestion
cmake-build-debug/
cmake-build-release/
```

**Comment by aminya** [comment]
@davidcampelo Could you resolve the suggestions that you have applied already?

**Comment by ddalcino** [comment]
Question regarding the first 16 lines: Are they intended for an in-source build? As far as I can tell, the only way for a lot of those files to get into your project root directory is if you built your project there. 

Since this project no longer supports in-source builds, should these lines still be in the .gitignore file?

**Comment by lefticus** [comment]
It's actually because of this PR, which I looked at then failed to comment on, that I disabled in-source builds

**Comment by Jason5480** [comment]
Shall we also ignore binary files in the source repo? The proposed method to use third party tools and executables is through conan packages. So I would add the following lines in .gitignore file
````
# Do not ever commit binaries in a source repo. Create a conan package instead!

# Compiled Object files
*.slo
*.lo
*.o
*.obj

# Precompiled Headers
*.gch
*.pch

# Compiled Dynamic libraries
*.so
*.dylib
*.dll

# Compiled Static libraries
*.lai
*.la
*.a
*.lib

# Executables
*.exe
*.out
*.app
````

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Makefile
+**CTestTestfile.cmake
+test/**.cmake
+
+# Conan files
+/conan.cmake
+conan.lock
+conanbuildinfo.cmake
+conanbuildinfo.txt
+conanfile.txt
```
Comment (aminya) [inline .gitignore:15]:
```suggestion
```

This conflicts with #65

**Comment by lefticus** [comment]
Because I have disabled in-source builds, I'm going to close this PR.

If you disagree, please reopen and comment.

**Comment by aminya** [comment]
@lefticus I think despite this, having gitignore does not have any disadvantageous. Someone might prefer to use in-source builds.

**Comment by ddalcino** [comment]
@lefticus I am a little disappointed that this didn't get merged in some form either. It's useful to have a `.gitignore` file here, even if it's empty. Its presence will remind users that they need to add things to it.

Personally, I would remove the first 18 lines, to make it doubly clear that in-source builds are not a good idea, and keep the rest. The rest of the file won't be useful for everyone, but I think parts of it are useful to most people.  None of the lines after 18 are likely to conflict with something you would want to commit in a C++ project, and if they do, it's not hard for a user to discover and remove them.

Also, I wanted to thank you for taking the time to maintain this repo. I recognize that it's an often thankless job, and most your interactions with users are with people like me who want you to read/evaluate/merge their PRs, or people in the 'Issues' section who are having trouble. I think we all owe you our thanks and a tall glass of beer.

**Pull request closed without merging**